### PR TITLE
[Inference/Refactor] Delete Duplicated code and refactor vec_copy utils and reduce utils

### PR DESCRIPTION
**Pull request by Courtesy-Xs**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs
- [ ] I have installed pre-commit: `pip install pre-commit && pre-commit install`

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?
This PR do the following things:
1. delete duplicated and unused scaled_softmax.py
2. delete type_converter which is used to trans torch type to cuda built-in vec type in fact
3. move vec_copy into colossalAI::cuda::utils
4. move reduce_max_into_lanes and so on to multi_tensor_l2norm because implementation of it is copied and seems that some bugs in it because it used uninitialized data in shared memory,  meanwhile, it was only used by multi_tensor_l2norm which owns no corresponding test case, most implementation for colossalAI training has no corresponding test case. maybe colleague who is responsible for train part will fix them in the future
5. complete cast functor

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

**Comment by Courtesy-Xs** [comment]
image

File: extensions/csrc/cuda/utils/vec_type_traits.h [hunk 1]
```diff
@@ -1,8 +1,9 @@
 #pragma once
 
-#include <c10/macros/Macros.h>
```
Comment (yuanheng-zhao) [inline extensions/csrc/cuda/utils/vec_type_traits.h:3]: You might want to remove `#include <c10/macros/Macros.h>` in `extensions/csrc/cuda/utils/vec_copy.h` as well.

File: extensions/csrc/cuda/multi_tensor_l2norm_kernel.cu [hunk 2]
```diff
@@ -12,14 +12,98 @@
...
-#include "include/block_reduce.h"
+#include "funcs/reduce_function.h"
 
 #define BLOCK_SIZE 512
 #define ILP 4
 
-using colossalAI::cuda::utils::block_reduce;
-using colossalAI::cuda::utils::reduce_block_into_lanes;
-using colossalAI::cuda::utils::reduce_block_into_lanes_max_op;
+
```
Comment (yuanheng-zhao) [inline extensions/csrc/cuda/multi_tensor_l2norm_kernel.cu:20]: Better to run relevant tests at training side.

**Comment by yuanheng-zhao** [comment]
Remember to edit the PR title to be reflecting your works within (concise and explicit summary). And please notice to check the commit title and messages when we do "Squash and merge".

File: extensions/csrc/cuda/utils/vec_type_traits.h (hunk 1 above)
Comment (Courtesy-Xs) [reply extensions/csrc/cuda/utils/vec_type_traits.h:3]: Done

**Comment by Courtesy-Xs** [comment]
image

File: extensions/csrc/cuda/multi_tensor_l2norm_kernel.cu (hunk 2 above)
Comment (Courtesy-Xs) [reply extensions/csrc/cuda/multi_tensor_l2norm_kernel.cu:20]: Done

**Review by yuanheng-zhao — APPROVED** [review]
LGTM

**Pull request merged**